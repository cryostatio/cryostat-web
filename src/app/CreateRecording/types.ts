/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EventTemplate, KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';

export type EventTemplateIdentifier = Pick<EventTemplate, 'name' | 'type'>;

interface _FormBaseData {
  name: string;
  template?: EventTemplateIdentifier;
  labels: KeyValue[];
  continuous: boolean;
  archiveOnStop: boolean;
  restart: boolean;
  duration: number;
  durationUnit: number;
  skipDurationCheck?: boolean;
  maxAge: number;
  maxAgeUnit: number;
  maxSize: number;
  maxSizeUnit: number;
  toDisk: boolean;
}

interface _FormValidationData {
  nameValid: ValidatedOptions;
  labelsValid: ValidatedOptions;
  durationValid: ValidatedOptions;
}

export type CustomRecordingFormData = _FormBaseData & _FormValidationData;
