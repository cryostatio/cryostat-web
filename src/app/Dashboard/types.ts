/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { FeatureLevel } from '@app/Shared/Services/service.types';
import { gridSpans, LabelProps } from '@patternfly/react-core';
import { Observable } from 'rxjs';

export interface Sized<T> {
  minimum: T;
  default: T;
  maximum: T;
}

export interface DashboardCardSizes {
  span: Sized<gridSpans>;
  height: Sized<number>;
}

export interface DashboardCardDescriptor {
  featureLevel: FeatureLevel;
  icon?: React.ReactNode;
  labels?: {
    content: string;
    color?: LabelProps['color'];
    icon?: React.ReactNode;
  }[];
  preview?: React.ReactNode;
  title: string;
  cardSizes: DashboardCardSizes;
  description: string;
  descriptionFull: JSX.Element | string;
  component: DashboardCardFC<DashboardCardTypeProps>;
  propControls: PropControl[];
  advancedConfig?: JSX.Element;
}
export type DashboardCardFC<P> = React.FC<P> & {
  cardComponentName: string;
};

export interface PropControlExtra {
  displayMapper?: (value: string) => string /* only has effect with 'select' PropControl kind */;
  min?: number;
  max?: number;
  [key: string]: unknown;
}

export interface PropControl {
  name: string;
  key: string;
  description: string;
  kind: 'boolean' | 'number' | 'string' | 'text' | 'select';
  values?: unknown[] | Observable<unknown>;
  defaultValue: unknown;
  extras?: PropControlExtra;
}

export interface DashboardCardTypeProps {
  span: number;
  dashboardId: number;
  isDraggable?: boolean;
  isResizable?: boolean;
  isFullHeight?: boolean;
  actions?: JSX.Element[];
}

export interface ValidationError {
  message: React.ReactNode;
}

export interface CardValidationResult {
  errors: ValidationError[];
  callForAction?: React.ReactNode;
}

export type LayoutTemplateFilter = 'Suggested' | 'Cryostat' | 'User-submitted';

export interface LayoutTemplateController {
  selectedTemplate: SelectedLayoutTemplate | undefined;
  setSelectedTemplate: (template: React.SetStateAction<SelectedLayoutTemplate | undefined>) => void;
  isUploadModalOpen: boolean;
  setIsUploadModalOpen: (isOpen: React.SetStateAction<boolean>) => void;
}

export interface CardConfig {
  id: string;
  name: string;
  span: gridSpans;
  props: object;
}

export type SerialCardConfig = Omit<CardConfig, 'id'>;

export const mockSerialCardConfig: SerialCardConfig = {
  name: 'Default',
  span: 12,
  props: {},
};
export interface DashboardLayout {
  name: string;
  cards: CardConfig[];
  favorite: boolean;
}

export type SerialDashboardLayout = Omit<DashboardLayout, 'name' | 'cards'> & { cards: SerialCardConfig[] };

// only name and vendor are needed to identify a template
export type LayoutTemplateRecord = Pick<LayoutTemplate, 'name' | 'vendor'>;

export enum LayoutTemplateVersion {
  'v2.3' = 'v2.3',
  'v2.4' = 'v2.4',
}

export enum LayoutTemplateVendor {
  BLANK = 'Blank',
  CRYOSTAT = 'Cryostat',
  USER = 'User-submitted',
}

export interface LayoutTemplate {
  name: string;
  description: string;
  cards: SerialCardConfig[];
  version: LayoutTemplateVersion;
  vendor: LayoutTemplateVendor;
}

export interface SelectedLayoutTemplate {
  template: LayoutTemplate;
  category: LayoutTemplateFilter;
}

export type SerialLayoutTemplate = Omit<LayoutTemplate, 'vendor'>;
