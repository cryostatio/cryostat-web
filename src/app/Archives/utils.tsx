/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RecordingDirectory, Target } from '@app/Shared/Services/api.types';

export const includesDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): boolean => {
  return arr.some((t) => t.connectUrl === dir.connectUrl);
};

export const indexOfDirectory = (arr: RecordingDirectory[], dir: RecordingDirectory): number => {
  let index = -1;
  arr.forEach((d, idx) => {
    if (d.connectUrl === dir.connectUrl) {
      index = idx;
    }
  });
  return index;
};

export const getTargetFromDirectory = (dir: RecordingDirectory): Target => {
  return {
    connectUrl: dir.connectUrl,
    alias: dir.jvmId,
    labels: [],
    annotations: {
      cryostat: [],
      platform: [],
    },
  };
};
