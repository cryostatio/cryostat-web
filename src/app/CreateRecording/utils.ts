/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TemplateType } from '@app/Shared/Services/api.types';
import { EventTemplateIdentifier } from './types';

export const RecordingNamePattern = /^[\w_]+$/;

export const DurationPattern = /^[1-9][0-9]*$/;

export const isRecordingNameValid = (name: string) => RecordingNamePattern.test(name);

export const isDurationValid = (duration: number) => DurationPattern.test(`${duration}`);

export const templateFromEventSpecifier = (eventSpecifier: string): EventTemplateIdentifier | undefined => {
  const regex = /^template=([a-zA-Z0-9]+)(?:,type=([a-zA-Z0-9]+))?$/im;
  if (eventSpecifier && regex.test(eventSpecifier)) {
    const parts = regex.exec(eventSpecifier);
    if (parts) {
      return {
        name: parts[1],
        type: parts[2] as TemplateType,
      };
    }
  }
  return undefined;
};
