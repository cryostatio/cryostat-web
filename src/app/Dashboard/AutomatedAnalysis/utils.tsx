/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AnalysisResult, Evaluation } from '@app/Shared/Services/api.types';
import { Stack, StackItem, Label, Title, Text } from '@patternfly/react-core';
import _ from 'lodash';

export const transformAADescription = (result: AnalysisResult): JSX.Element => {
  const format = (s: Evaluation): JSX.Element => {
    if (typeof s === 'string') {
      return <Text>{s}</Text>;
    }
    if (Array.isArray(s)) {
      return (
        <Stack>
          {s.map((e) => (
            <StackItem key={e.setting}>
              <Title headingLevel={'h6'}>{e.setting}</Title>
              <Label>
                {e.name}={e.value}
              </Label>
            </StackItem>
          ))}
        </Stack>
      );
    }
    throw `Unrecognized item: ${s}`;
  };
  return (
    <div>
      {Object.entries(result.evaluation || {}).map(([k, v]) =>
        v && v.length ? (
          <div key={k}>
            <span>
              <Title headingLevel={'h5'}>{_.capitalize(k)}</Title>
              {format(result.evaluation[k])}
            </span>
            <br />
          </div>
        ) : (
          <div key={k}></div>
        ),
      )}
    </div>
  );
};
