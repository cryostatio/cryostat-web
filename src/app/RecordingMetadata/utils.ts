/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { KeyValue } from '@app/Shared/Services/api.types';
import { ValidatedOptions } from '@patternfly/react-core';
import { Observable, from } from 'rxjs';

export const isEqualLabel = (a: KeyValue, b: KeyValue) => {
  return a.key === b.key && a.value === b.value;
};

export const includesLabel = (arr: KeyValue[], searchLabel: KeyValue) => {
  return arr.some((l) => isEqualLabel(searchLabel, l));
};

export const parseLabelsFromFile = (file: File): Observable<KeyValue[]> => {
  return from(
    file
      .text()
      .then(JSON.parse)
      .then((obj) => {
        const labels: KeyValue[] = [];
        const labelObj = obj['labels'];
        if (labelObj) {
          Object.keys(labelObj).forEach((key) => {
            labels.push({
              key: key,
              value: labelObj[key],
            });
          });
          return labels;
        }
        throw new Error('No labels found in file');
      }),
  );
};

export const getLabelDisplay = (label: KeyValue) => `${label.key}:${label.value}`;

export const LabelPattern = /^\S+$/;

export const getValidatedOption = (isValid: boolean) => {
  return isValid ? ValidatedOptions.success : ValidatedOptions.error;
};

export const matchesLabelSyntax = (l: KeyValue) => {
  return l && LabelPattern.test(l.key) && LabelPattern.test(l.value);
};
