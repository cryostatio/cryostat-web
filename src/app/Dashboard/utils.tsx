/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import cryostatLogo from '@app/assets/cryostat_icon_rgb_default.svg';
import cryostatLogoDark from '@app/assets/cryostat_icon_rgb_reverse.svg';
import { dashboardConfigDeleteCardIntent } from '@app/Shared/Redux/ReduxStore';
import { FeatureLevel } from '@app/Shared/Services/service.types';
import { withThemedIcon } from '@app/utils/withThemedIcon';
import { useCryostatTranslation } from '@i18n/i18nextUtil';
import { Button, ButtonVariant } from '@patternfly/react-core';
import { FileIcon, UnknownIcon, UserIcon } from '@patternfly/react-icons';
import { nanoid } from '@reduxjs/toolkit';
import { TFunction } from 'i18next';
import * as React from 'react';
import { useDispatch } from 'react-redux';
import { AutomatedAnalysisCardDescriptor } from './AutomatedAnalysis/AutomatedAnalysisCard';
import { JFRMetricsChartCardDescriptor } from './Charts/jfr/JFRMetricsChartCard';
import { MBeanMetricsChartCardDescriptor } from './Charts/mbean/MBeanMetricsChartCard';
import { DiagnosticsCardDescriptor } from './Diagnostics/DiagnosticsCard';
import { JvmDetailsCardDescriptor } from './JvmDetails/JvmDetailsCard';
import {
  SerialLayoutTemplate,
  SerialCardConfig,
  LayoutTemplateVersion,
  LayoutTemplateController,
  LayoutTemplateVendor,
  DashboardLayout,
  LayoutTemplate,
  LayoutTemplateRecord,
  CardConfig,
  CardValidationResult,
  ValidationError,
  DashboardCardDescriptor,
} from './types';

export const DashboardLayoutNamePattern = /^[a-zA-Z0-9_.-]+( [a-zA-Z0-9_.-]+)*$/;
export const LayoutTemplateDescriptionPattern = /^[a-zA-Z0-9\s.,\-'";?!@#$%^&*()[\]_+=:{}]*$/;

export const mockSerialLayoutTemplate: SerialLayoutTemplate = {
  name: 'Default',
  description: 'Default.',
  cards: [] as SerialCardConfig[],
  version: LayoutTemplateVersion['v2.3'],
};

// use a provider
export const LayoutTemplateContext = React.createContext<LayoutTemplateController>({
  selectedTemplate: undefined,
  setSelectedTemplate: () => undefined,
  isUploadModalOpen: false,
  setIsUploadModalOpen: () => undefined,
});

export const iconify = (vendor: LayoutTemplateVendor): React.ReactNode => {
  switch (vendor) {
    case LayoutTemplateVendor.CRYOSTAT:
      return <ThemedCryostatLogo />;
    case LayoutTemplateVendor.BLANK:
      return <FileIcon style={{ paddingRight: '0.3rem' }} />;
    case LayoutTemplateVendor.USER:
      return <UserIcon />;
    default:
      return <UnknownIcon />;
  }
};

export const ThemedCryostatLogo = withThemedIcon(cryostatLogo, cryostatLogoDark, 'Cryostat Logo');

export const templatize = (layout: DashboardLayout, name: string, desc?: string): LayoutTemplate => {
  return {
    name: name,
    description: desc || `Custom layout template.`,
    cards: layout.cards.map((card) => {
      const { id: _id, ...cardWithoutId } = card;
      return { ...cardWithoutId };
    }) as SerialCardConfig[],
    vendor: LayoutTemplateVendor.USER,
    version: LayoutTemplateVersion['v2.3'],
  } as LayoutTemplate;
};

export const layoutize = (template: LayoutTemplate, name: string): DashboardLayout => {
  return {
    name: name,
    cards: template.cards.map((card) => {
      return {
        ...card,
        id: `${card.name}-${nanoid()}`,
      };
    }),
    favorite: false,
  } as DashboardLayout;
};

export const recordToLayoutTemplate = (
  record: LayoutTemplateRecord,
  allTemplates: LayoutTemplate[],
): LayoutTemplate | undefined => {
  return allTemplates.find((template) => template.name === record.name && template.vendor === record.vendor);
};

export const getUniqueIncrementingName = (init = 'Custom', names: string[]): string => {
  let name: string;
  let i = 1;
  do {
    name = `${init}${i}`;
    i++;
  } while (names.includes(name));
  return name;
};

export const hasCardDescriptorByName = (name: string): boolean => {
  for (const choice of getDashboardCards()) {
    if (choice.component.cardComponentName === name) {
      return true;
    }
  }
  return false;
};

export const getCardDescriptorByName = (name: string): DashboardCardDescriptor => {
  for (const choice of getDashboardCards()) {
    if (choice.component.cardComponentName === name) {
      return choice;
    }
  }
  throw new Error(`Unknown card type selection: ${name}`);
};

export const hasCardDescriptorByTitle = (title: string, t: TFunction): boolean => {
  for (const choice of getDashboardCards()) {
    if (t(choice.title) === title) {
      return true;
    }
  }
  return false;
};

export const getCardDescriptorByTitle = (title: string, t: TFunction): DashboardCardDescriptor => {
  for (const choice of getDashboardCards()) {
    if (t(choice.title) === title) {
      return choice;
    }
  }
  throw new Error(`Unknown card type selection: ${title}`);
};

export const getDashboardCards: (featureLevel?: FeatureLevel) => DashboardCardDescriptor[] = (
  featureLevel = FeatureLevel.DEVELOPMENT,
) => {
  const cards = [
    JvmDetailsCardDescriptor,
    AutomatedAnalysisCardDescriptor,
    JFRMetricsChartCardDescriptor,
    MBeanMetricsChartCardDescriptor,
    DiagnosticsCardDescriptor,
  ];
  return cards.filter((card) => card.featureLevel >= featureLevel);
};

export const RemoveCardAction: React.FC<{ cardIndex: number }> = ({ cardIndex }) => {
  const dispatch = useDispatch();
  const { t } = useCryostatTranslation();

  const handleClick = React.useCallback(() => {
    dispatch(dashboardConfigDeleteCardIntent(cardIndex));
  }, [dispatch, cardIndex]);

  return (
    <Button onClick={handleClick} variant={ButtonVariant.danger}>
      {t('REMOVE')}
    </Button>
  );
};

export const validateCardConfig = ({ name, props }: CardConfig, cardIndex: number): CardValidationResult => {
  // Unsupported card type
  if (!hasCardDescriptorByName(name)) {
    return {
      errors: [
        {
          message: (
            <>
              Unknown card type: <code>{name}</code>.
            </>
          ),
        },
      ],
      callForAction: <RemoveCardAction cardIndex={cardIndex} />,
    };
  }

  const errs: ValidationError[] = [];

  const { propControls } = getCardDescriptorByName(name);
  const configPropKeys = Object.entries(props);

  // Missing props
  propControls
    .map((ctrl) => ctrl.key)
    .forEach((propKey) => {
      const matched = configPropKeys.find(([key, _]) => key === propKey);
      if (!matched) {
        errs.push({
          message: (
            <>
              Missing card property: <code>{propKey}</code>.
            </>
          ),
        });
      }
    });

  // Invalid prop valies
  configPropKeys.forEach(([propKey, propValue]) => {
    const matched = propControls.find((crtl) => crtl.key === propKey);
    if (matched) {
      const { values, extras, kind } = matched;
      // FIXME: Check dynamic values
      let err: ValidationError | undefined;
      if (propValue === undefined || propValue === null) {
        err = {
          message: (
            <>
              Undefined value for card property: <code>{propKey}</code>
            </>
          ),
        };
      } else if (values && Array.isArray(values) && !values.some((v) => v === propValue)) {
        err = {
          message: (
            <>
              Invalid value <code>{propValue}</code> for card property: <code>{propKey}</code>.
            </>
          ),
        };
      } else if (extras) {
        if (kind === 'number') {
          const valAsNum = Number(propValue);
          if (isNaN(valAsNum)) {
            err = {
              message: (
                <>
                  Numeric value expected but <code>{typeof propValue}</code> was given for: <>{propKey}</>.
                </>
              ),
            };
          } else {
            if (
              (extras.min !== undefined && valAsNum < extras.min) ||
              (extras.max !== undefined && valAsNum > extras.max)
            ) {
              err = {
                message: (
                  <>
                    Value exceeding limit for card property: <code>{propKey}</code>.
                  </>
                ),
              };
            }
          }
        }
      }
      err && errs.push(err);
    }
    // Unknown props are ignored
  });

  return {
    errors: errs,
    callForAction: errs.length ? <RemoveCardAction cardIndex={cardIndex} /> : undefined,
  };
};
