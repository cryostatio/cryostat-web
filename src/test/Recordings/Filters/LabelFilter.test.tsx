/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { LabelFilter } from '@app/Recordings/Filters/LabelFilter';
import { ActiveRecording, RecordingState } from '@app/Shared/Services/api.types';
import { cleanup, screen, within } from '@testing-library/react';
import { render, renderSnapshot } from '../../utils';

const mockRecordingLabels = [
  {
    key: 'someLabel',
    value: 'someValue',
  },
];
const mockAnotherRecordingLabels = [
  {
    key: 'anotherLabel',
    value: 'anotherValue',
  },
];
const mockRecordingLabelList = ['someLabel:someValue', 'anotherLabel:anotherValue'];

const mockRecording: ActiveRecording = {
  name: 'someRecording',
  downloadUrl: 'http://downloadUrl',
  reportUrl: 'http://reportUrl',
  metadata: { labels: mockRecordingLabels },
  startTime: 1234567890,
  id: 0,
  state: RecordingState.RUNNING,
  duration: 0,
  continuous: false,
  toDisk: false,
  maxSize: 0,
  maxAge: 0,
};
const mockAnotherRecording = {
  ...mockRecording,
  name: 'anotherRecording',
  metadata: { labels: mockAnotherRecordingLabels },
} as ActiveRecording;
const mockRecordingWithoutLabel = {
  ...mockRecording,
  name: 'noLabelRecording',
  metadata: { labels: [] },
} as ActiveRecording;
const mockRecordingList = [mockRecording, mockAnotherRecording, mockRecordingWithoutLabel];

const onLabelInput = jest.fn((_label) => {
  /**Do nothing. Used for checking renders */
});

describe('<LabelFilter />', () => {
  let emptyFilteredLabels: string[];
  let filteredLabels: string[];

  beforeEach(() => {
    emptyFilteredLabels = [];
    filteredLabels = [`someLabel:someValue`];
  });

  afterEach(cleanup);

  it('renders correctly', async () => {
    const tree = await renderSnapshot({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: (
              <LabelFilter
                recordings={mockRecordingList}
                filteredLabels={emptyFilteredLabels}
                onSubmit={onLabelInput}
              />
            ),
          },
        ],
      },
    });
    expect(tree?.toJSON()).toMatchSnapshot();
  });

  it('display label selections when text input is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: (
              <LabelFilter
                recordings={mockRecordingList}
                onSubmit={onLabelInput}
                filteredLabels={emptyFilteredLabels}
              />
            ),
          },
        ],
      },
    });
    const labelInput = screen.getByLabelText('Filter by label...');
    expect(labelInput).toBeInTheDocument();
    expect(labelInput).toBeVisible();

    await user.click(labelInput);

    const selectMenu = await screen.findByRole('listbox', { name: 'Filter by label' });
    expect(selectMenu).toBeInTheDocument();
    expect(selectMenu).toBeVisible();

    mockRecordingLabelList.forEach((label) => {
      const option = within(selectMenu).getByText(label);
      expect(option).toBeInTheDocument();
      expect(option).toBeVisible();
    });
  });

  it('display label selections when dropdown arrow is clicked', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: (
              <LabelFilter
                recordings={mockRecordingList}
                onSubmit={onLabelInput}
                filteredLabels={emptyFilteredLabels}
              />
            ),
          },
        ],
      },
    });

    const dropDownArrow = screen.getByRole('button', { name: 'Options menu' });
    expect(dropDownArrow).toBeInTheDocument();
    expect(dropDownArrow).toBeVisible();

    await user.click(dropDownArrow);

    const selectMenu = await screen.findByRole('listbox', { name: 'Filter by label' });
    expect(selectMenu).toBeInTheDocument();
    expect(selectMenu).toBeVisible();

    mockRecordingLabelList.forEach((label) => {
      const option = within(selectMenu).getByText(label);
      expect(option).toBeInTheDocument();
      expect(option).toBeVisible();
    });
  });

  it('should close selection menu when toggled with dropdown arrow', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: (
              <LabelFilter
                recordings={mockRecordingList}
                onSubmit={onLabelInput}
                filteredLabels={emptyFilteredLabels}
              />
            ),
          },
        ],
      },
    });

    const dropDownArrow = screen.getByRole('button', { name: 'Options menu' });
    expect(dropDownArrow).toBeInTheDocument();
    expect(dropDownArrow).toBeVisible();

    await user.click(dropDownArrow);

    const selectMenu = await screen.findByRole('listbox', { name: 'Filter by label' });
    expect(selectMenu).toBeInTheDocument();
    expect(selectMenu).toBeVisible();

    mockRecordingLabelList.forEach((label) => {
      const option = within(selectMenu).getByText(label);
      expect(option).toBeInTheDocument();
      expect(option).toBeVisible();
    });

    await user.click(dropDownArrow);
    expect(selectMenu).not.toBeInTheDocument();
    expect(selectMenu).not.toBeVisible();
  });

  it('should close selection menu when toggled with text input', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: (
              <LabelFilter
                recordings={mockRecordingList}
                onSubmit={onLabelInput}
                filteredLabels={emptyFilteredLabels}
              />
            ),
          },
        ],
      },
    });
    const labelInput = screen.getByLabelText('Filter by label...');
    expect(labelInput).toBeInTheDocument();
    expect(labelInput).toBeVisible();

    await user.click(labelInput);

    const selectMenu = await screen.findByRole('listbox', { name: 'Filter by label' });
    expect(selectMenu).toBeInTheDocument();
    expect(selectMenu).toBeVisible();

    mockRecordingLabelList.forEach((label) => {
      const option = within(selectMenu).getByText(label);
      expect(option).toBeInTheDocument();
      expect(option).toBeVisible();
    });

    await user.click(labelInput);
    expect(selectMenu).not.toBeInTheDocument();
    expect(selectMenu).not.toBeVisible();
  });

  it('should not display selected labels', async () => {
    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: (
              <LabelFilter recordings={mockRecordingList} onSubmit={onLabelInput} filteredLabels={filteredLabels} />
            ),
          },
        ],
      },
    });
    const labelInput = screen.getByLabelText('Filter by label...');
    expect(labelInput).toBeInTheDocument();
    expect(labelInput).toBeVisible();

    await user.click(labelInput);

    const selectMenu = await screen.findByRole('listbox', { name: 'Filter by label' });
    expect(selectMenu).toBeInTheDocument();
    expect(selectMenu).toBeVisible();

    const notToShowLabel = within(selectMenu).queryByText('someLabel:someValue');
    expect(notToShowLabel).not.toBeInTheDocument();
  });

  it('should select a name when a name option is clicked', async () => {
    const submitLabelInput = jest.fn((labelInput) => emptyFilteredLabels.push(labelInput));

    const { user } = render({
      routerConfigs: {
        routes: [
          {
            path: '/recordings',
            element: (
              <LabelFilter
                recordings={mockRecordingList}
                onSubmit={submitLabelInput}
                filteredLabels={emptyFilteredLabels}
              />
            ),
          },
        ],
      },
    });
    const labelInput = screen.getByLabelText('Filter by label...');
    expect(labelInput).toBeInTheDocument();
    expect(labelInput).toBeVisible();

    await user.click(labelInput);

    const selectMenu = await screen.findByRole('listbox', { name: 'Filter by label' });
    expect(selectMenu).toBeInTheDocument();
    expect(selectMenu).toBeVisible();

    mockRecordingLabelList.forEach((label) => {
      const option = within(selectMenu).getByText(label);
      expect(option).toBeInTheDocument();
      expect(option).toBeVisible();
    });

    await user.click(screen.getByText('someLabel:someValue'));

    // NameFilter's parent rebuilds to close menu by default.

    expect(submitLabelInput).toBeCalledTimes(1);
    expect(submitLabelInput).toBeCalledWith('someLabel:someValue');
    expect(emptyFilteredLabels).toStrictEqual(['someLabel:someValue']);
  });
});
