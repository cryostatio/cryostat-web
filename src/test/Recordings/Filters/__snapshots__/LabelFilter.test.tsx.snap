// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LabelFilter /> renders correctly 1`] = `
<div
  className="pf-v5-c-menu-toggle pf-m-full-width pf-m-typeahead"
>
  <div
    className="pf-v5-c-text-input-group pf-m-plain"
  >
    <div
      autoComplete="off"
      className="pf-v5-c-text-input-group__main"
      id="typeahead-label-filter"
      onClick={[Function]}
    >
      <span
        className="pf-v5-c-text-input-group__text"
      >
        <input
          aria-controls="typeahead-label-select"
          aria-expanded={false}
          aria-label="Type to filter"
          className="pf-v5-c-text-input-group__text-input"
          onChange={[Function]}
          placeholder="Filter by label..."
          role="combobox"
          type="text"
          value=""
        />
      </span>
    </div>
    <div
      className="pf-v5-c-text-input-group__utilities"
    />
  </div>
  <button
    aria-expanded={false}
    aria-label="Menu toggle"
    className="pf-v5-c-menu-toggle__button"
    onClick={[Function]}
    type="button"
  >
    <span
      className="pf-v5-c-menu-toggle__controls"
    >
      <span
        className="pf-v5-c-menu-toggle__toggle-icon"
      >
        <svg
          aria-hidden={true}
          aria-labelledby={null}
          className="pf-v5-svg"
          fill="currentColor"
          height="1em"
          role="img"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          />
        </svg>
      </span>
    </span>
  </button>
</div>
`;
